doc/slarchive.1
.TH SLARCHIVE 1 2010/03/10
.SH NAME
slarchive \- SeedLink client for data stream archiving

.SH SYNOPSIS
.nf
slarchive [options] [host][:][port]

.fi
.SH DESCRIPTION
\fBslarchive\fP connects to a \fISeedLink\fR server, requests data
streams and writes received packets into directory/file structures.
The precise layout of the directories and files is defined in a
\fIformat\fP string.  Some preset file layouts are included: the
SeisComP Data Structure (SDS), Buffer of Uniform Data structure (BUD)
and a few others.  To write more than one archive simply specify
multiple \fIformat\fP definitions (or presets).

.SH OPTIONS

.IP "-V         "
Print program version and exit.

.IP "-h         "
Print program usage and exit.

.IP "-H         "
Print program usage including details for the replacement flags used
with the \fI-A\fP option and exit.

.IP "-v         "
Be more verbose.  This flag can be used multiple times ("-v -v" or
"-vv") for more verbosity.  One flag: report basic handshaking (link
configuration) details and briefly report each packet received.  Two
flags: report the details of the handshaking, each packet received and
detailed connection diagnostics.

.IP "-p             "
Print details of received miniSEED data records. This flag can be
used multiple times ("-p -p" or "-pp") for more detail.  One flag: a
single summary line for each data packet received.  Two flags: details
of the miniSEED data records received, including information from
fixed header and 100/1000/1001 blockettes.

.IP "-f \fImax\fR"
The maximum number of files to keep open.  If not specified the
maximum will be the default system process limit (for the given
environment).  The maxium number of open archive data files will be
maintained at 10 less than the absolute maximum leaving descriptors
available for other tasks (e.g. writing state files).

.IP "-nd \fIdelay\fR"
The network reconnect delay (in seconds) for the connection to
the SeedLink server.  If the connection breaks for any reason
this will govern how soon a reconnection should be attempted.
The default value is 30 seconds.

.IP "-nt \fItimeout\fR"
The network timeout (in seconds) for the connection to the SeedLink
server.  If no data [or keep alive packets?] are received in this
time the connection is closed and re-established (after the
reconnect delay has expired).  The default value is 600 seconds.
A value of 0 disables the timeout.

.IP "-k \fIkeepalive\fR  (requires SeedLink >= 3)"
Keepalive packet interval (in seconds) at which keepalive (heartbeat)
packets are sent to the server.  Keepalive packets are only sent if
nothing is received within the interval.

.IP "-x \fIstatefile\fR[:\fIinterval\fR]"
During client shutdown the last received sequence numbers and time
stamps (start times) for each data stream will be saved in this file.
If this file exists upon startup the information will be used to
resume the data streams from the point at which they were stopped.  In
this way the client can be stopped and started without data loss,
assuming the data are still available on the server.  If
\fIinterval\fR is specified the state will be saved every
\fIinterval\fR packets that are received.  Otherwise the state
will be saved only on normal program termination.

.IP "-i \fItimeout\fR"
Timeout for closing idle data stream files in seconds.  The idle time
of data streams is only checked when a packet has arrived so if no
packets are arriving no idle stream files will be closed.  There
should be no reason to change this parameter except for unusual cases
where the process is running against an open file number limit.
Default is 300 seconds.

.IP "-d"
Configure the connection in "dial-up" mode.  The remote server will
close the connection when it has sent all of the data in it's buffers
for the selected data streams.  This is opposed to the normal behavior
of waiting indefinately for data.

.IP "-b"
Configure the connection in "batch" mode.  Negotiation with the remote
server is made faster by minimizing acknowledgement checks.

.IP "-Fi[:overlap]  "
Future check initially.  Check the last miniSEED data record in an
existing archive file and do not write new data to that file if it is
older within a certain overlap.  The default overlap limit is 2
seconds; the overlap can be specified by appending a colon and the
desired overlap limit in seconds to the option.  If the overlap is
exceeded an error message will be logged once for each time the file
is opened.  This option only makes sense for archive formats where
each unique data stream is written to a unique file (e.g. SDS format).
If a data stream is closed due to timeout (see option -i) the initial
future check will be preformed again if the file is re-opened.

.IP "-Fc[:overlap]  "
Future check continuously.  Only archive miniSEED data records if the
first sample of the record is newer, within a certain overlap, than
the last sample of the previous record for a given archive file.  The
default overlap limit is 2 seconds; the overlap can be specified by
appending a colon and the desired overlap limit in seconds to the
option.  If the overlap is exceeded an error message will be logged
once until either a non-overlapping packet is received or a new
archive file is used.  This option only makes sense for archive
formats where each unique data stream is written to a unique file
(e.g. SDS format).

.IP "-s \fIselectors\fR"
This defines default selectors.  If no multi-station data streams are
configured these selectors will be used for uni/all-station mode.
Otherwise these selectors will be used when no selectors are specified
for a given stream using the '-S' or '-l' options.

.IP "-l \fIlistfile\fR"
A list of streams will be read from the given file.  This option
implies multi-station mode.  The format of the stream list file is
given below in the section \fIStream list file\fR.

.IP "-S \fIstream[:selectors],...\fR  (requires SeedLink >= 2.5)"
The connection will be configured in multi-station mode with optional
SeedLink selectors for each station, see examples below.  \fIstream\fR
should be provided in NET_STA format.  If supported by the server
the '*' or '?' wildcards may be used in the NET_STA designation.  If
no selectors are provided for a given stream, the default selectors,
if defined, will be used.

.IP "-tw \fIstart:[end]\fR  (requires SeedLink >= 3)"
Specifies a time window applied, by the server, to data streams.  The
format for both times is year,month,day,hour,min,sec; for example:
"2002,08,05,14,00:2002,08,05,14,15,00".  The end time is optional but
the colon must be present.  If no end time is specified the server
will send data indefinately.  This option will override any saved
state information.

Warning: time windowing might be disabled on the remote server.

.IP "-A \fIformat\fR"
If specified, all packets (miniSEED records) received will be
appended to a directory/file structure defined by \fBformat\fP.  All
directories implied in the \fBformat\fP string will be created if
necessary.  The option may be used multiple times to write received
packets to multiple archives.  See the section \fIArchiving data\fR.

.IP "-SDS \fISDSdir\fR"
If specified, all packets (miniSEED records) received will be saved
into a Simple Data Structure (SDS) dir/file structure starting at the
specified directory.  This directory and all subdirectories will be
created if necessary.  This option is a preset version of the '-A'
option.  The SDS dir/file structure is:
.nf
<SDSdir>/<YEAR>/<NET>/<STA>/<CHAN.TYPE>/NET.STA.LOC.CHAN.TYPE.YEAR.DAY
.fi

.IP "-BUD \fIBUDdir\fR"
If specified, all waveform data packets (miniSEED data records)
received will be saved into a Buffer of Uniform Data (BUD) dir/file
structure starting at the specified directory.  This directory and all
subdirectories will be created if necessary.  This option is a preset
version of the '-A' option.  The BUD dir/file structure is:
.nf
<BUDdir>/<NET>/<STA>/STA.NET.LOC.CHAN.YEAR.DAY
.fi

.IP "-CSS \fICSSdir\fR"
If specified, all packets (miniSEED data records) received will be
saved into a CSS-like data file structure starting at the specified
directory.  This directory and all subdirectories will be created if
necessary.  This option is a preset version of the '-A' option.  The
CSS dir/file structure is:
.nf
<CSSdir>/<YEAR>/<DAY>/STA.CHAN.YEAR:DAY:HOUR:MIN:SEC
.fi
where the HOUR, MIN & SEC fields are non-defining substitutions.

.IP "-CHAN \fIdir\fR"
If specified, all packets (miniSEED data records) received will be
saved into channel files in the specified directory.  This directory
will be created if necessary.  This option is a preset version of
the '-A' option.  The CHAN dir/file structure is:
.nf
<dir>/NET.STA.LOC.CHAN
.fi

.IP "-QCHAN \fIdir\fR"
If specified, all packets (miniSEED data records) received will be
saved into channel files including the data quality identifier in the
specified directory.  This directory will be created if necessary.
This option is a preset version of the '-A' option.  The QCHAN dir/file
structure is:
.nf
<dir>/NET.STA.LOC.CHAN.QUALITY
.fi

.IP "-CDAY \fIdir\fR"
If specified, all packets (miniSEED data records) received will be
saved into channel-day files in the specified directory.  This
directory will be created if necessary.  This option is a preset
version of the '-A' option.  The CDAY dir/file structure is:
.nf
<dir>/NET.STA.LOC.CHAN.YEAR.DAY:HOUR:MIN:SEC
.fi
where the HOUR, MIN & SEC fields are non-defining substitutions.

.IP "\fI[host][:][port]\fR"
A required argument, specifies the address of the SeedLink server in
host:port format.  Either the host, port or both can be omitted.  If
host is omitted then localhost is assumed, i.e.  ':18000'
implies 'localhost:18000'.  If the port is omitted then 18000 is
assumed, i.e.  'localhost' implies 'localhost:18000'.  If only ':' is
specified 'localhost:18000' is assumed.

.SH "SeedLink SELECTORS"

ChangeLog
2023.051: 3.2
	- Update libslink to 2.7.1.
	- Fix crash on record parsing error
	- Do not skip so-called end-of-detection records, where sample rate
	factor is 0 and number of samples is 0.

2023.006: 3.1
	- Update libslink to 2.7.
	- Fix minor compiler warning on printf().

2016.295:
	- Update libslink to 2.6.

2013.316: 3.0
	- Properly handle the case where all output files associated with a
	DataStream would be closed when the idle time out reached 0 by making
	sure the entry currently being used does not get closed.  The previous
	logic allowed all files to be closed if they were all opened within
	the last second since the timer has a resolution of one second.
	- Only check for and create parent directories when opening a file,
	instead of checking for parent directories for each record written.
	This change eliminates some unnecessary system I/O calls, since there
	is no need to check for parent directories for files already open.
	- Create directories and files with full permissions: directories (777)
	and files (666).  An operator can control the final permissions of the
	files and directories by setting the umask.
	- Cleanup some unused, extra variables.

2013.305: 2.2
	- Update to libslink 2.4b with fix for negating selectors in stream
	selection list file.

2010.102: 2.1
	- Fix globbing detection and bad filename pointer use,
	reported by Eric Haug.

2010.075:
	- Update to libslink 2.3 with support for batch mode.
	- Add support for batch mode, patch by Andres Heinloo/GFZ.

2009.318: 2.0b
	- Update to libslink 2.2c, with minor fixes.

2008.113: 2.0
	- Update to libslink 2.2 (inconsequential changes).

2008.043: 2.0 beta3
	- Add logic to retry interrupted write(2) calls up to 10
	times for interrupted calls to guard against partial writes.

2008.029: 2.0 beta2
	- Update to libslink 2.1.
	- Change src/Makefile to use user defined CC and CFLAGS.
	- Add option -f to specify the max number of open files.
	- Change ds_openfile() to maintain the number of open files
	to be less than the maximum number allowed minus 10, thus
	leaving 10 descriptors for other use.

2008.024: 2.0 beta
	- Update to libslink 2.0 with support for network/station wildcards.
	- Resynchronize archiving routines with updates from other
	code bases, eliminate archive.[ch] source files.
	- Add 'q', 'L', 'r' and 'R' archive formatting flags.
	- Remove DLOG preformatted archive type.
	- Add CSS, CHAN, QCHAN and CDAY preformatted archive types.
	- Update man page.

2007.183:
	- Improved error handling, mostly in the case of interrupted
	system calls (due to signals) in dsarchive.c.

2006.213: 1.7
	- Release without changes after testing.

2005.237:
	- Search for a matching file name (glob) when opening archive
	files using non-defining flags.  The problem was that stream
        timeouts or slarchive re-starts cause files to be re-opened
	generating new files using the non-defining flags as if they were
	defining.

2005.147: version 1.6
	- Update to libslink 1.5.
	- Add version to usage messages.
	- Update man page.

2004.222: version 1.5
	- Update to libslink 1.4.
	- Either host or port is optional when specifying the
        SeedLink server (libslink parses the string now).
	- Remove the dummy signal handler as it was useless.

2004.197: version 1.4
	- Update to libslink 1.3 including redirection of most library
	output to stderr.
	- Do not adjust ppackets flag based on verbose flag.  If packet
        details are desired they must specifically be requested.

2004.184:
	- Migrate all standard I/O (fopen, fwrite, etc.) to the raw
	interfaces (open, write, etc.) in order to be able to open more
	files at once; standard I/O is limited to 256 descriptors.
	- Agressively deal with the situation where the maximum open files
	limit has been reached.  First, try to increase the "soft" limit
	to the "hard" limit.  Second, close open stream files with
	decreasing idle timeouts until some file descriptors are freed.
	The result is that slarchive should almost always be able to open
	files (and save data) even if the number of incoming streams is
	large.  If the open file limit is low versus the number of
	incoming streams there might be lots of opening and closing of
	files.  This change involved the addition of the ds_openfile and
	ds_closeidle functions to dsarchive.c.
	- If '-v' is used only set '-p' to level 1 unless it's already
	set, this allows diagnostics to be seen without the output being
	flooded with info about the received packets.

2004.170: version 1.3
	- Update to libslink 1.2a including more flexible support for
        implementations of the SeedLink protocol other than the reference
        implementation by Andres Heinloo.
	- Handle case of corrupt miniSEED records in existing files when
	checking for the last sample received (initial future check) by
	disabling check when the last sample time could not be determined.

src/slarchive.c
/***************************************************************************
 * slarchive.c
 *
 * A SeedLink client for data collection and archiving.
 *
 * Connects to a SeedLink server, collects and archives data.
 *
 * Written by Chad Trabant
 *   ORFEUS/EC-Project MEREDIAN
 *   IRIS Data Management Center
 *   EarthScope Data Services
 ***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>

#include <libslink.h>

#include "dsarchive.h"

#define PACKAGE   "slarchive"
#define VERSION   "3.2"

static void packet_handler (char *msrecord, int packet_type, int seqnum);
static int  parameter_proc (int argcount, char **argvec);
static char *getoptval (int argcount, char **argvec, int argopt);
static int  addarchive(const char *path, const char *layout);
static void term_handler (int sig);
static void print_timelog (const char *msg);
static void usage (int level);

/* A chain of archive definitions */
typedef struct DSArchive_s {
  DataStream  datastream;
  struct DSArchive_s *next;
}
DSArchive;

static short int verbose  = 0;   /* flag to control general verbosity */
static short int ppackets = 0;   /* flag to control printing of data packets */
static int stateint       = 0;   /* packet interval to save statefile */
static char *statefile    = 0;	 /* state file for saving/restoring stream states */

static SLCD *slconn;	         /* connection parameters */
static DSArchive *dsarchive;

int
main (int argc, char **argv)
{
  SLpacket *slpack;
  int seqnum;
  int ptype;
  int packetcnt = 0;

  /* Signal handling, use POSIX calls with standardized semantics */
  struct sigaction sa;

  sa.sa_flags   = SA_RESTART;
  sigemptyset (&sa.sa_mask);

  sa.sa_handler = term_handler;
  sigaction (SIGINT, &sa, NULL);
  sigaction (SIGQUIT, &sa, NULL);
  sigaction (SIGTERM, &sa, NULL);

  sa.sa_handler = SIG_IGN;
  sigaction (SIGHUP, &sa, NULL);
  sigaction (SIGPIPE, &sa, NULL);

  dsarchive = NULL;

  /* Allocate and initialize a new connection description */
  slconn = sl_newslcd();

  /* Process given parameters (command line and parameter file) */
  if (parameter_proc (argc, argv) < 0)
    {
      fprintf (stderr, "Parameter processing failed\n");
      fprintf (stderr, "Try '-h' for detailed help\n");
      return -1;
    }

  /* Loop with the connection manager */
  while ( sl_collect (slconn, &slpack) )
    {
      ptype  = sl_packettype (slpack);
      seqnum = sl_sequence (slpack);

      packet_handler (slpack->msrecord, ptype, seqnum);

      if ( statefile && stateint )
	{
	  if ( ++packetcnt >= stateint )
	    {
	      sl_savestate (slconn, statefile);
	      packetcnt = 0;
	    }
	}
    }

  /* Do all the necessary cleanup and exit */
  if (slconn->link != -1)
    sl_disconnect (slconn);

  if (dsarchive) {
    DSArchive *curdsa = dsarchive;

    while ( curdsa != NULL ) {
      ds_streamproc (&curdsa->datastream, NULL, 0);
      curdsa = curdsa->next;
    }
  }

  if (statefile)
    sl_savestate (slconn, statefile);

  return 0;
}  /* End of main() */


/***************************************************************************
 * packet_handler:
 * Process a received packet based on packet type.
 ***************************************************************************/
static void
packet_handler (char *msrecord, int packet_type, int seqnum)
{
  static SLMSrecord * msr = NULL;

  double dtime;			/* Epoch time */
  double secfrac;		/* Fractional part of epoch time */
  time_t ttime;			/* Integer part of epoch time */
  char timestamp[36] = {0};
  struct tm *timep;
  int    archflag = 1;

  /* The following is dependent on the packet type values in libslink.h */
  char *type[]  = { "Data", "Detection", "Calibration", "Timing",
		    "Message", "General", "Request", "Info",
                    "Info (terminated)", "KeepAlive" };

  if (verbose >= 1)
  {
    /* Build a current local time string */
    dtime   = sl_dtime ();
    secfrac = (double) ((double)dtime - (int)dtime);
    ttime   = (time_t) dtime;
    timep   = localtime (&ttime);
    snprintf (timestamp, sizeof(timestamp), "%04d.%03d.%02d:%02d:%02d.%01.0f",
	      timep->tm_year + 1900, timep->tm_yday + 1, timep->tm_hour,
	      timep->tm_min, timep->tm_sec, secfrac);

    sl_log (1, 1, "%s, seq %d, Received %s blockette\n",
	    timestamp, seqnum, type[packet_type]);
  }

  /* Parse data record and print requested detail if any */
  sl_msr_parse (slconn->log, msrecord, &msr, 1, 0);

  if (msr == NULL)
  {
    sl_log (2, 0, "cannot parse miniSEED record\n");
    return;
  }

  if ( ppackets )
    sl_msr_print (slconn->log, msr, ppackets - 1);

  /* Write packet to all archives in archive definition chain */
  if (dsarchive && archflag)
  {
    DSArchive *curdsa = dsarchive;

    while ( curdsa != NULL ) {
      curdsa->datastream.packettype = packet_type;

      ds_streamproc (&curdsa->datastream, msr, 0);

      curdsa = curdsa->next;
    }
  }
}  /* End of packet_handler() */


/***************************************************************************
 * parameter_proc:
 *
 * Process the command line parameters.
 *
 * Returns 0 on success, and -1 on failure.
 ***************************************************************************/
static int
parameter_proc (int argcount, char **argvec)
{
  int futurecontflag= 0;  /* continuous future check flag */
  int futurecont    = 2;  /* continuous future check overlap */
  int futureinitflag= 0;  /* initial future check (opening files) flag */
  int futureinit    = 2;  /* initial future check (opening files) overlap */
  int idletimeout   = 300; /* idle stream timeout */
  int error = 0;

  char *streamfile  = 0;   /* stream list file for configuring streams */
  char *multiselect = 0;
  char *selectors   = 0;
  char *timewin     = 0;
  char *tptr;

  SLstrlist *timelist;	   /* split the time window arg */

  if (argcount <= 1)
    error++;

  /* Process all command line arguments */
  for (optind = 1; optind < argcount; optind++)
    {
      if (strcmp (argvec[optind], "-V") == 0)
        {
          fprintf (stderr, "%s version: %s\n", PACKAGE, VERSION);
          exit (0);
        }
      else if (strcmp (argvec[optind], "-h") == 0)
        {
          usage (0);
          exit (0);
        }
      else if (strcmp (argvec[optind], "-H") == 0)
        {
          usage (1);
          exit (0);
        }
      else if (strncmp (argvec[optind], "-v", 2) == 0)
	{
	  verbose += strspn (&argvec[optind][1], "v");
	}
      else if (strncmp (argvec[optind], "-p", 2) == 0)
        {
          ppackets += strspn (&argvec[optind][1], "p");
        }
      else if (strncmp (argvec[optind], "-Fc", 3) == 0)
        {
	  futurecontflag = 1;
	  if ( (tptr = strchr(argvec[optind], ':')) )
	    futurecont = atoi(tptr+1);
        }
      else if (strncmp (argvec[optind], "-Fi", 3) == 0)
        {
	  futureinitflag = 1;
	  if ( (tptr = strchr(argvec[optind], ':')) )
	    futureinit = atoi(tptr+1);
        }
      else if (strcmp (argvec[optind], "-d") == 0)
	{
	  slconn->dialup = 1;
	}
      else if (strcmp (argvec[optind], "-b") == 0)
	{
	  slconn->batchmode = 1;
	}
      else if (strcmp (argvec[optind], "-f") == 0)
	{
	  ds_maxopenfiles = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-nd") == 0)
	{
	  slconn->netdly = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-nt") == 0)
	{
	  slconn->netto = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-k") == 0)
	{
	  slconn->keepalive = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-i") == 0)
	{
	  idletimeout = atoi (getoptval(argcount, argvec, optind++));
	}
      else if (strcmp (argvec[optind], "-A") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), NULL) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-SDS") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), SDSLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-BUD") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), BUDLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-CSS") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), CSSLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-CHAN") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), CHANLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-QCHAN") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), QCHANLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-CDAY") == 0)
	{
	  if ( addarchive(getoptval(argcount, argvec, optind++), CDAYLAYOUT) == -1 )
	    return -1;
	}
      else if (strcmp (argvec[optind], "-l") == 0)
	{
	  streamfile = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-s") == 0)
	{
	  selectors = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-S") == 0)
	{
	  multiselect = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-x") == 0)
	{
	  statefile = getoptval(argcount, argvec, optind++);
	}
      else if (strcmp (argvec[optind], "-tw") == 0)
	{
	  timewin = getoptval(argcount, argvec, optind++);
	}
      else if (strncmp (argvec[optind], "-", 1) == 0)
	{
	  fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
	  exit (1);
	}
      else if (!slconn->sladdr)
        {
          slconn->sladdr = argvec[optind];
        }
      else
	{
	  fprintf (stderr, "Unknown option: %s\n", argvec[optind]);
	  exit (1);
	}
    }

  /* Make sure a server was specified */
  if ( ! slconn->sladdr )
    {
      fprintf(stderr, "No SeedLink server specified\n\n");
      fprintf(stderr, "%s version %s\n\n", PACKAGE, VERSION);
      fprintf(stderr, "Usage: %s [options] [host][:][port]\n\n", PACKAGE);
      fprintf(stderr, "Try '-h' for detailed help\n");
      exit (1);
    }

  /* Initialize the verbosity for the sl_log function */
  sl_loginit (verbose, &print_timelog, NULL, &print_timelog, NULL);

  /* Set stdout (where logs go) to always flush after a newline */
  setvbuf(stdout, NULL, _IOLBF, 0);

  /* Report the program version */
  sl_log (1, 1, "%s version: %s\n", PACKAGE, VERSION);

  /* If errors then report the usage message and quit */
  if ( error )
    {
      usage (0);
      exit (1);
    }

  /* Load the stream list from a file if specified */
  if ( streamfile )
    sl_read_streamlist (slconn, streamfile, selectors);

  /* Split the time window argument */
  if ( timewin )
    {

      if (strchr (timewin, ':') == NULL)
	{
	  sl_log (2, 0, "time window not in begin:[end] format\n");
	  return -1;
	}

      if (sl_strparse (timewin, ":", &timelist) > 2)
	{
	  sl_log (2, 0, "time window not in begin:[end] format\n");
	  sl_strparse (NULL, NULL, &timelist);
	  return -1;
	}

      if (strlen (timelist->element) == 0)
	{
	  sl_log (2, 0, "time window must specify a begin time\n");
	  sl_strparse (NULL, NULL, &timelist);
	  return -1;
	}

      slconn->begin_time = strdup (timelist->element);

      timelist = timelist->next;

      if (timelist != 0)
	{
	  slconn->end_time = strdup (timelist->element);

	  if (timelist->next != 0)
	    {
	      sl_log (2, 0, "malformed time window specification\n");
	      sl_strparse (NULL, NULL, &timelist);
	      return -1;

	    }
	}

      /* Free the parsed list */
      sl_strparse (NULL, NULL, &timelist);
    }

  /* Parse the 'multiselect' string following '-S' */
  if ( multiselect )
    {
      if ( sl_parse_streamlist (slconn, multiselect, selectors) == -1 )
	return -1;
    }
  else if ( !streamfile )
    {		         /* No 'streams' array, assuming uni-station mode */
      sl_setuniparams (slconn, selectors, -1, 0);
    }

  /* Attempt to recover sequence numbers from state file */
  if (statefile)
    {
      /* Check if interval was specified for state saving */
      if ((tptr = strchr (statefile, ':')) != NULL)
	{
	  char *tail;

	  *tptr++ = '\0';

	  stateint = (unsigned int) strtoul (tptr, &tail, 0);

	  if ( *tail || (stateint < 0 || stateint > 1e9) )
	    {
	      sl_log (2, 0, "state saving interval specified incorrectly\n");
	      return -1;
	    }
	}

      if (sl_recoverstate (slconn, statefile) < 0)
	{
	  sl_log (2, 0, "state recovery failed\n");
	}
    }

  /* If no archiving is specified print a warning */
  if ( !dsarchive ) {
    sl_log (1, 0, "WARNING: no archiving method was specified\n");
  }
  /* Otherwise fill in the global parameters for each entry */
  else {
    DSArchive *curdsa = dsarchive;
    while ( curdsa != NULL ) {
      curdsa->datastream.idletimeout = idletimeout;
      curdsa->datastream.futurecontflag = futurecontflag;
      curdsa->datastream.futurecont = futurecont;
      curdsa->datastream.futureinitflag = futureinitflag;
      curdsa->datastream.futureinit = futureinit;
      curdsa = curdsa->next;
    }
  }

  return 0;
}  /* End of parameter_proc() */


/***************************************************************************
 * getoptval:
 *
 * Return the value to a command line option; checking that the value is
 * itself not an option (starting with '-') and is not past the end of
 * the argument list.
 *
 * argcount: total arguments in argvec
 * argvec: argument list
 * argopt: index of option to process, value is expected to be at argopt+1
 *
 * Returns value on success and exits with error message on failure
 ***************************************************************************/
static char *
getoptval (int argcount, char **argvec, int argopt)
{
  if ( argvec == NULL || argvec[argopt] == NULL ) {
    fprintf (stderr, "getoptval(): NULL option requested\n");
    exit (1);
  }

  if ( (argopt+1) < argcount && *argvec[argopt+1] != '-' )
    return argvec[argopt+1];

  fprintf (stderr, "Option %s requires a value\n", argvec[argopt]);
  exit (1);
}  /* End of getoptval() */


/***************************************************************************
 * addarchive():
 *
 * Add entry to the data stream archive chain.  If 'layout' is defined
 * it will be appended to 'path'.
 *
 * Returns 0 on success, and -1 on failure
 ***************************************************************************/
static int
addarchive ( const char *path, const char *layout )
{
  DSArchive *newdsa;
  int pathlayout;

  if ( ! path )
    {
      sl_log (2, 0, "addarchive(): cannot add archive with empty path\n");
      return -1;
    }

  if ( ! (newdsa = (DSArchive *) malloc (sizeof (DSArchive))) )
    {
      sl_log (2, 0, "cannot allocate memory for new archive definition\n");
      return -1;
    }

  /* Setup new entry and add it to the front of the chain */
  pathlayout = strlen (path) + 2;
  if ( layout )
    pathlayout += strlen (layout);

  if ( ! (newdsa->datastream.path = (char *) malloc (pathlayout)) )
    {
      sl_log (2, 0, "cannot allocate memory for new archive path\n");
      if ( newdsa )
	free (newdsa);
      return -1;
    }

  if ( layout )
    snprintf (newdsa->datastream.path, pathlayout, "%s/%s", path, layout);
  else
    snprintf (newdsa->datastream.path, pathlayout, "%s", path);

  newdsa->datastream.grouproot = NULL;

  newdsa->next = dsarchive;
  dsarchive = newdsa;

  return 0;
}  /* End of addarchive() */


/***************************************************************************
 * term_handler:
 * Signal handler routine to controll termination.
 ***************************************************************************/
static void
term_handler (int sig)
{
  sl_terminate (slconn);
}


/***************************************************************************
 * print_timelog:
 * Log message print handler used with sl_loginit() and sl_log().  Prefixes
 * a local time string to the message before printing.
 ***************************************************************************/
static void
print_timelog (const char *msg)
{
  char timestr[100];
  time_t loc_time;

  /* Build local time string and cut off the newline */
  time(&loc_time);
  strcpy(timestr, asctime(localtime(&loc_time)));
  timestr[strlen(timestr) - 1] = '\0';

  fprintf (stdout, "%s - %s", timestr, msg);
}


/***************************************************************************
 * usage:
 * Print the usage message and exit.
 ***************************************************************************/
static void
usage (int level)
{
  fprintf (stderr, "%s version %s\n\n", PACKAGE, VERSION);
  fprintf (stderr, "Usage: %s [options] [host][:][port]\n\n", PACKAGE);
  fprintf (stderr,
	   " ## General program options ##\n"
	   " -V              Report program version\n"
	   " -h              Print this usage message\n"
           " -H              Print usage message with 'format' details (see -A option)\n"
	   " -v              Be more verbose, multiple flags can be used\n"
	   " -p              Print details of data packets, multiple flags can be used\n"
	   " -f max          Maximum number of files to keep open\n"
	   " -nd delay       Network re-connect delay (seconds), default 30\n"
	   " -nt timeout     Network timeout (seconds), re-establish connection if no\n"
	   "                   data/keepalives are received in this time, default 600\n"
	   " -k interval     Send keepalive (heartbeat) packets this often (seconds)\n"
	   " -x sfile[:int]  Save/restore stream state information to this file\n"
	   " -i timeout      Idle stream entries might be closed (seconds), default 300\n"
	   " -d              Configure the connection in dial-up mode\n"
	   " -b              Configure the connection in batch mode\n"
	   " -Fi[:overlap]   Initially check (existing files) that data records are newer\n"
	   " -Fc[:overlap]   Continuously check that data records are newer\n"
	   "\n"
	   " ## Data stream selection ##\n"
	   " -s selectors    Selectors for uni/all-station or default for multi-station mode\n"
	   " -l listfile     Read a stream list from this file for multi-station mode\n"
           " -S streams      Define a stream list for multi-station mode\n"
	   "   'streams' = 'stream1[:selectors1],stream2[:selectors2],...'\n"
	   "        'stream' is in NET_STA format, for example:\n"
	   "        -S \"IU_KONO:BHE BHN,GE_WLF,MN_AQU:HH?.D\"\n\n"
	   " -tw begin:[end]  (requires SeedLink >= 3)\n"
	   "        specify a time window in year,month,day,hour,min,sec format\n"
	   "        example: -tw 2002,08,05,14,00,00:2002,08,05,14,15,00\n"
	   "        the end time is optional, but the colon must be present\n"
	   "\n"
	   " ## Data archiving options ##\n"
	   " -A format       Write records is a custom file structure (try -H)\n"
	   " -SDS SDSdir     Write records in a SDS file structure\n"
	   " -BUD BUDdir     Write records in a BUD file layout\n"
	   " -CSS CSSdir     Write records in a CSS-like file layout\n"
	   " -CHAN dir       Write records into separate Net.Sta.Loc.Chan files\n"
	   " -QCHAN dir      Write records into separate Net.Sta.Loc.Chan.Quality files\n"
	   " -CDAY dir       Write records into separate Net.Sta.Loc.Chan-day files\n"
	   "\n"
	   " [host][:][port] Address of the SeedLink server in host:port format\n"
	   "                   Default host is 'localhost' and default port is '18000'\n\n");

  if  ( level )
    {
      fprintf (stderr,
               "The archive 'format' argument is expanded for each record using the\n"
               "following flags:\n"
               "\n"
               "  n : Network code, white space removed\n"
               "  s : Station code, white space removed\n"
               "  l : Location code, white space removed\n"
               "  c : Channel code, white space removed\n"
               "  Y : Year, 4 digits\n"
               "  y : Year, 2 digits zero padded\n"
               "  j : Day of year, 3 digits zero padded\n"
               "  H : Hour, 2 digits zero padded\n"
               "  M : Minute, 2 digits zero padded\n"
               "  S : Second, 2 digits zero padded\n"
               "  F : Fractional seconds, 4 digits zero padded\n"
               "  q : record quality indicator (D, R, Q, M), single character\n"
               "  L : Data record length in bytes\n"
               "  r : Sample rate (Hz) as a rounded integer\n"
               "  R : Sample rate (Hz) as a float with 6 digit precision\n"
               "  t : SeedLink packet type, single character\n"
               "  %% : The percent (%%) character\n"
               "  # : The number (#) character\n"
               "\n"
               "The flags are prefaced with either the %% or # modifier.  The %% modifier\n"
               "indicates a defining flag while the # indicates a non-defining flag.\n"
               "All records with the same set of defining flags will be written to the\n"
               "same file. Non-defining flags will be expanded using the values in the\n"
               "first record for the resulting file name.\n"
               "\n");
    }
}  /* End of usage() */
